### PR TITLE
Replace the internal Fibonacci heap with a 4-ary priority queue

### PR DESCRIPTION
**Pull request by Jake-Moss**
This PR provides up to a 3.5x performance increase on the various models tested.
It also changes the default behaviour of `memory_only` mode to default to `True`, enables the `NPY_NO_DEPRECATED_API` and `NPY_1_7_API_VERSION` macros for Unix builds, and restores the `language_level` to `3` for Cython.

The heap implementation were written by @djfrancesco and has the MIT license preserved as a header in the `pq_4ary_heap.pyx` file.

### CODE REVIEW

File: aequilibrae/paths/basic_path_finding.pyx [hunk 1]
```diff
@@ -14,7 +14,7 @@
  Licence:     See LICENSE.TXT
  -----------------------------------------------------------------------------------------------------------
 Original Algorithm for Shortest path (Dijkstra with a Fibonacci heap) was written by Jake Vanderplas <vanderplas@astro.washington.edu> under license: BSD, (C) 2012
```
Comment (jamiecook) [inline aequilibrae/paths/basic_path_finding.pyx:16]: probably need to change this.

Comment (Jake-Moss) [reply aequilibrae/paths/basic_path_finding.pyx:16]: Good catch, I've updated the other banner as well

**Review by djfrancesco — APPROVED** [review]

**Review by djfrancesco — APPROVED** [review]

**Review by pedrocamargo — APPROVED** [review]

**Comment by pedrocamargo** [comment]
For reference. Image created by @chrisobi02 
MicrosoftTeams-image

**Comment by pedrocamargo** [comment]
Congrats to @Jake-Moss , @chrisobi02 & @djfrancesco for the amazing work that went into this PR!!!

**Pull request merged**